### PR TITLE
Add error handling

### LINKED ISSUE #71: Add error handling to API-bound things
* Redirect to oauth flow on auth error

Show error page on error for:
* GET /linodes
* GET /distributions

Show error response inline (overall error and/or per-field error) for:
* Alerts form
* Edit group / label

Show error tooltip for:
* Boot jobs
  * Bulk reboot (and power off, etc.) on /linodes
  * Single reboot on /linodes
  * Single reboot on /linodes/:id

### PR DESCRIPTION
**Pull request by SirCmpwn**
* [x] Redirect to oauth flow on auth error
* [ ] Tests

Show error page on error for:
* [x] GET /linodes
* [x] GET /linodes/:id
* [x] GET /distributions
* [x] Handle 404 errors appropriately
* [ ] Handle uncaught exceptions

Show error response inline (overall error and/or per-field error) for:
* [x] Alerts form
* [x] Edit group / label

Closes #71

### CODE REVIEW

**Comment by SirCmpwn** [comment]
Anyone can display the error page by using this pattern:

```js
try {
  await dispatch(fetchLinodes());
} catch (response) {
  dispatch(setError(response));
}
```

**Comment by eatonphil** [comment]
That's pretty nice!

**Comment by eatonphil** [comment]
Maybe retry instead of refresh though.

File: src/linodes/layouts/CreateLinodePage.js [hunk 1]
```diff
@@ -8,16 +8,23 @@ import OrderSummary from '../components/OrderSummary';
...
-    dispatch(fetchServices());
+    try {
+      await Promise.all([
+        dispatch(fetchDistros()),
+        dispatch(fetchDatacenters()),
+        dispatch(fetchServices()),
+      ]);
+    } catch (response) {
+      dispatch(setError(response));
+    }
```
Comment (SirCmpwn) [inline src/linodes/layouts/CreateLinodePage.js]: @eatonphil fyi

Comment (eatonphil) [inline src/linodes/layouts/CreateLinodePage.js]: kk

**Comment by eatonphil** [comment]
It's a good start!

**Comment by SirCmpwn** [comment]
Let's hear your feedback.

**Comment by eatonphil** [comment]
I think it could probably use a bit of fiddling sometime (nothing concrete yet) but overall it's pretty good!

**Comment by SirCmpwn** [comment]
Alright, I'll leave the fiddling to you when this PR is done.

**Comment by eatonphil** [comment]
We'll probably need a refactor PR later putting this into a Form and uniting error handling with Form.

**Comment by SirCmpwn** [comment]
Turns out there aren't any validation errors that the alert page could generate from the API. Ticking that box.

**Comment by SirCmpwn** [comment]
Two problems:

1. Reboot doesn't raise any errors when the Linode can't be rebooted (i.e. no configs). I have filed bug 544 with the API team regarding this. Will open a Github issue for us to add error handling for it once it actually returns errors.
2. The power on button doesn't work. I think this is covered by #204.

So I'm just going to move boot-related error handling out of the scope of this PR.

File: src/layouts/Layout.js [hunk 2]
```diff
@@ -31,7 +31,7 @@ export class Layout extends Component {
             style={{ margin: '0 0.51rem' }}
             className="btn btn-default"
             onClick={() => window.location.reload(true)}
-          >Refresh</button>
+          >Reload</button>
```
Comment (eatonphil) [inline src/layouts/Layout.js]: ty

**Comment by SirCmpwn** [comment]
I'm okay with -0.2%. This PR adds some changes that are impractical to test.

File: src/actions/errors.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+export const SET_ERROR = '@@errors/SET_ERROR';
+export const TOGGLE_DETAILS = '@@errors/TOGGLE_DETAILS';
+
+export function setError(response) {
+  return async (dispatch) => {
+    const type = response.headers.get('Content-Type');
+    let json = 'fubar';
```
Comment (eatonphil) [inline src/actions/errors.js]: `let json = {}`?

File: src/linodes/actions/detail/index.js [hunk 4]
```diff
@@ -18,20 +19,48 @@ export function setLinodeGroup(group) {
   return { type: SET_LINODE_GROUP, group };
 }
 
+export function clearErrors() {
+  return { type: SET_ERRORS, label: null, group: null, _: null };
```
Comment (eatonphil) [inline src/linodes/actions/detail/index.js]: What is the _ here?

File: src/actions/errors.js (hunk 3 above)
Comment (SirCmpwn) [inline src/actions/errors.js]: No, this is intentional. It's part of the path where we don't get JSON from the server, instead perhaps we got an nginx error page. I don't want it to look like legit JSON in this case (we get {} from successful API responses sometimes, too).

File: src/linodes/actions/detail/index.js (hunk 4 above)
Comment (SirCmpwn) [inline src/linodes/actions/detail/index.js]: Errors that aren't associated with a particular field.

File: test/linodes/layouts/IndexPage.spec.js [hunk 5]
```diff
@@ -52,7 +53,29 @@ describe('linodes/layouts/IndexPage', () => {
     expect(fetchStub.calledOnce).to.equal(true);
     expect(fetchStub.firstCall.args[1]).to.equal('/linodes?page=1');
     expect(dispatch.calledOnce).to.equal(true);
-    expect(dispatch.firstCall.args[0].type).to.equal(UPDATE_LINODES);
+    expect(dispatch.firstCall.args[0].type).to.equal(linodeActions.UPDATE_LINODES);
+  });
+
+  it('handles errors from fetchLinodes', async () => {
```
Comment (eatonphil) [inline test/linodes/layouts/IndexPage.spec.js]: This async shouldn't be necessary.

File: test/linodes/layouts/LinodeDetailPage.spec.js [hunk 6]
```diff
@@ -58,7 +59,28 @@ describe('linodes/layouts/LinodeDetailPage/loadLinode', async () => {
     expect(fetchStub.calledOnce).to.equal(true);
     expect(fetchStub.firstCall.args[1]).to.equal('/linodes/linode_1234');
     expect(dispatch.calledOnce).to.equal(true);
-    expect(dispatch.firstCall.args[0].type).to.equal(UPDATE_LINODE);
+    expect(dispatch.firstCall.args[0].type).to.equal(linodeActions.UPDATE_LINODE);
+  });
+
+  it('handles errors from updateLinode', async () => {
```
Comment (eatonphil) [inline test/linodes/layouts/LinodeDetailPage.spec.js]: Neither should this?

**Comment by eatonphil** [comment]
Cool, looks good. Ready for merge?

**Comment by SirCmpwn** [comment]
Yes.

**Comment by eatonphil** [comment]
Just waiting for the tests passing then

**Comment by SirCmpwn** [comment]
Thanks!

**Pull request merged**